src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py

```diff
@@ -0,0 +1,235 @@
+        if meta["memory_mb"]:
+            memory = int(meta["memory_mb"]) - 2000 # keep 2gb for OS
+            cwl_data = re.sub('"ramMin": [^\n]*[^,](,?)\n', f'"ramMin": {memory}\\1\n', cwl_data)
+        if meta["cpus"]:
+            cwl_data = re.sub('"coresMin": [^\n]*[^,](,?)\n', f'"coresMin": {meta["cpus"]}\\1\n', cwl_data)
+
+        # Write the file out again
+        with open(cwl_file, 'w') as file:
+            file.write(cwl_data)
+    else:
+        cwl_file = orig_cwl_file
+
+    return os.path.abspath(cwl_file)
+
+def copy_outputs(par: dict[str, Any], config: dict[str, Any]):
+    for arg in config["arguments"]:
+        par_value = par[arg["clean_name"]]
+        if par_value and arg["type"] == "file" and arg["direction"] == "output":
+            # example template: '[sample_name]_(assay)_cell_type_experimental.csv'
+            template = (arg.get("info") or {}).get("template")
+            if template:
+                template_glob = template\
+                    .replace("[sample_name]", par["run_name"])\
+                    .replace("(assay)", "*")\
+                    .replace("[number]", "*")
+                files = glob.glob(os.path.join(par["output_dir"], template_glob))
+                if len(files) == 0 and arg["required"]:
+                    raise ValueError(f"Expected output file '{template_glob}' not found.")
+                elif len(files) > 1 and not arg["multiple"]:
+                    raise ValueError(f"Expected single output file '{template_glob}', but found multiple.")
+                
+                if not arg["multiple"]:
+                    shutil.copy(files[0], par_value)
+                else:
+                    # replace '*' in par_value with index
+                    for i, file in enumerate(files):
+                        shutil.copy(file, par_value.replace("*", str(i)))
+
+
+def main(par: dict[str, Any], meta: dict[str, Any], temp_dir: str):
+    config = read_config(meta["config"])
+    
+    # Preprocess params
+    par = process_params(par, config, temp_dir)
+
+    ## Process parameters
+    cmd = [
+        "cwl-runner",
+        "--no-container",
+        "--preserve-entire-environment",
+        "--outdir", par["output_dir"],
+    ]
+
+    if par["parallel"]:
+        cmd.append("--parallel")
+
+    if par["timestamps"]:
+        cmd.append("--timestamps")
+
+    # Create cwl file (if need be)
+    cwl_file = generate_cwl_file(meta, temp_dir)
+    cmd.append(cwl_file)
+
+    # Create params file
+    config_file = generate_config_file(par, config, temp_dir)
+    cmd.append(config_file)
+    
+    # keep environment variables but set TMPDIR to temp_dir
+    env = dict(os.environ)
+    env["TMPDIR"] = temp_dir
+
+    # Create output dir if not exists
+    if not os.path.exists(par["output_dir"]):
+        os.makedirs(par["output_dir"])
+
+    # Run command
+    print("> " + ' '.join(cmd), flush=True)
+    _ = subprocess.check_call(
```

Comment:
Do you mean `out = subprocess.check_output(cmd_pars).decode("utf-8")` and then printing the stdout?

Reply:
Yes, printing it, or writing it to an output file.

Reply:
I want to view the output as it comes out of the process ^^

Reply:
Printing it is also fine. But right now, neither one of those is happening (because check_call is used). https://docs.python.org/3/library/subprocess.html#subprocess.check_call

```
Code needing to capture stdout or stderr should use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) instead:
```

Reply:
I don't understand. `subprocess.check_call` does exactly what I want it to do.

```python
>>> out = subprocess.check_call(["echo", "hi"])
hi
```

For sure, `subprocess.run` seems to do exactly the same:

```python
>>> out = subprocess.run(["echo", "hi"], check=True)
hi
```

I guess I don't understand what the difference between the two variants is and why you'd like me to use `subprocess.run` over `subprocess.check_call`.

Reply:
Turns out that the pipeline already returns its own log in the output, so this has become a non-issue 👍